File: repos/koi-backend/flutter_lib/lib/storage/repositories/task_repository.dart
import 'package:hive/hive.dart';
import '../hive_manager.dart';
import '../../models/hive_task_model.dart';

class TaskRepository {
  Box<HiveTask> get _box => HiveManager.tasks;

  Future<void> saveTask(HiveTask task) async => _box.put(task.id, task);

  HiveTask? getTask(String id) => _box.get(id);

  List<HiveTask> getActiveByUser(String userId) {
    return _box.values
        .where((t) => t.userId == userId && !t.archived)
        .toList()
      ..sort((a, b) => a.createdAt.compareTo(b.createdAt));
  }

  List<HiveTask> getCompletedByUser(String userId) {
    return _box.values
        .where((t) => t.userId == userId && t.completedAt != null)
        .toList()
      ..sort((a, b) => a.createdAt.compareTo(b.createdAt));
  }

  List<HiveTask> getUnsynced() =>
      _box.values.where((t) => !t.synced).toList();

  Future<void> markSynced(String id) async {
    final task = _box.get(id);
    if (task == null) return;
    task.synced = true;
    await task.save();
  }

  Future<void> completeTask(String id) async {
    final task = _box.get(id);
    if (task == null) return;
    task.completedAt = DateTime.now();
    await task.save();
  }

  Future<void> archiveTask(String id) async {
    final task = _box.get(id);
    if (task == null) return;
    task.archived = true;
    await task.save();
  }

  Future<void> deleteTask(String id) async => _box.delete(id);
}

File: repos/koi-web/src/lib/repositories/task_repository.ts
import { db } from '../db';
import type { Task } from '../../models/index';

export class TaskRepository {
  async save(task: Task): Promise<void> {
    await db.tasks.put(task);
  }

  async getById(id: string): Promise<Task | undefined> {
    return db.tasks.get(id);
  }

  async getActiveByUser(userId: string): Promise<Task[]> {
    return db.tasks
      .where('userId')
      .equals(userId)
      .filter((t) => !t.archived)
      .toArray();
  }

  async getCompletedByUser(userId: string): Promise<Task[]> {
    return db.tasks
      .where('userId')
      .equals(userId)
      .filter((t) => t.completedAt !== undefined)
      .toArray();
  }

  async getUnsynced(): Promise<Task[]> {
    return db.tasks.where('synced').equals(0).toArray();
  }

  async markSynced(id: string): Promise<void> {
    await db.tasks.update(id, { synced: true });
  }

  async complete(id: string): Promise<void> {
    await db.tasks.update(id, { completedAt: new Date() });
  }

  async archive(id: string): Promise<void> {
    await db.tasks.update(id, { archived: true });
  }

  async delete(id: string): Promise<void> {
    await db.tasks.delete(id);
  }
}

export const taskRepository = new TaskRepository();
